### DISCUSSION TITLE
Aliased=True deprecation, migration

### DISCUSSION THREAD

**Discussion by mmssix** [sqlalchemy/sqlalchemy#10764] [Usage Questions]
Evening.  Having an issue porting over 1.4 code.  I've hit a wall and hoping someone might be able to help.

I'm trying to update this [flask/datatables](https://github.com/orf/datatables/blob/master/datatables/__init__.py#L57) code to work under 2.0.  The line referenced in that link needs to be updated to use aliases but I have no idea how.  I did read [the migration notes](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#orm-query-join-aliased-true-from-joinpoint-removed), but I cant figure out how to apply that fix to this code.

A snippet of the code that causes the issue:
```python
class Users(db.Model):
    __tablename__ = 'app_users'
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(64), index=True, unique=True)
    first = db.Column(db.String(254))
    last = db.Column(db.String(254))

class Chats(db.Model):
    __tablename__ = "chat_chats"
    id = db.Column(db.Integer, primary_key=True)
    chat_name = db.Column(db.String(254), index=True)
    created_by_id = db.Column(db.Integer, db.ForeignKey('app_users.id'), nullable=False, index=True)
    created_by = db.relationship('Users', foreign_keys=[created_by_id])


table = DataTable(request.args, Chats, db.session.query(Chats),
    [
        "id",
        "chat_name",
        "created_at",
        ('created_by_name', 'created_by.last', lambda i: f"{i.first[0]}.{i.last}"), # <--- Causes the issue.
    ])
```

If the `created_by_name` tuple is removed in the `table` declaration, the error disappears.  If you put in a print around line 57 in the linked code, it prints the string `created_by.last`, so `column.model_name.split(".")[0]` would be `created_by`.

If i cant get this working, I'll have to rewrite a truly staggering amount of code for an existing codebase at work.  Any help would be hugely appreciated.  I would ask in the repository for that library but it seems to be well and truly dead.

**Comment by zzzeek** [comment]
Hey there -

well unfortunately it looks like this library is relying upon an API feature that we deprecated about four years ago and is removed in SQLAlchemy 2.0 - the part where they join to a string attribute name can be fixed on their end easily, by using `getattr()` to get the real attribute, but for the `aliased=True` part they would need to create an `aliased()` object for the target element and then join to that.   This will allow that code to work fine, however the one thing that would not work, and im not sure if datatables is doing this, would be if you are trying to add additional query filter criteria on that related element, that would also have to be done in terms of that aliased class.

Alternatively datatables here could be modified to not use aliased=True at all in that spot, but then that would affect the ability to set up joins to the same entity more than once.   Perhaps your application doesn't need that feature, however.

overall datatables looks like a tiny one-file application that's only 200 lines of code so certainly you can just vendor this code and make the few small modifications to have it work for your use case.  it's license seems to be MIT licensed so you can just use it.

**Reply by mmssix** [reply]
Sir,
I appreciate the response.  I cant believe I didnt think about `getattr()` myself.  I'll see what I can do with that.

As for just "vendoring the code", thats were it all falls down.  Im not knowledgeable enough yet to make the "few small modifications" needed.  I'm a self taught one-man-band at work trying to bring a legacy codebase into modern times.  Hence why I was asking here.  ;)

If it comes down to it, can you recommend a person or group that would be capable of re-writing this on commission?  I'm happy to throw a bit of money at it to save myself days of time at work.

Much appreciated.
